crates/ruff_python_ast/generate.py

```diff
@@ -0,0 +1,581 @@
+#!/usr/bin/python
+
+from __future__ import annotations
+from dataclasses import dataclass
+from pathlib import Path
+import re
+from subprocess import check_output
+import tomllib
+from typing import Optional
+
+
+def rustfmt(code: str) -> str:
+    return check_output(["rustfmt", "--emit=stdout"], input=code, text=True)
+
+
+def to_snake_case(node: str) -> str:
+    """Converts CamelCase to snake_case"""
+    return re.sub("([A-Z])", r"_\1", node).lower().lstrip("_")
+
+
+# ------------------------------------------------------------------------------
+# Read AST description
+
+root = check_output(["git", "rev-parse", "--show-toplevel"], text=True).strip()
+ast_path = Path(root).joinpath("crates", "ruff_python_ast", "ast.toml")
+out_path = Path(root).joinpath("crates", "ruff_python_ast", "src", "generated.rs")
+with ast_path.open("rb") as ast_file:
+    ast = tomllib.load(ast_file)
+
+# ------------------------------------------------------------------------------
+# Preprocess
+
+
+@dataclass
+class Group:
+    name: str
+    nodes: list[Node]
+    owned_enum_ty: str
+    ref_enum_ty: str
+
+    add_suffix_to_is_methods: bool
+    anynode_is_label: str
+    comment: Optional[str]
+
+    def __init__(self, group_name, group):
+        self.name = group_name
+        self.owned_enum_ty = group_name
+        self.ref_enum_ty = group.get("ref_enum_ty", group_name + "Ref")
+        self.add_suffix_to_is_methods = group.get("add_suffix_to_is_methods", False)
+        self.anynode_is_label = group.get("anynode_is_label", to_snake_case(group_name))
+        self.comment = group.get("comment")
+        self.nodes = [
+            Node(self, node_name, node) for node_name, node in group["nodes"].items()
+        ]
+
+
+@dataclass
+class Node:
+    name: str
+    variant: str
+    ty: str
+
+    def __init__(self, group, node_name, node):
+        self.name = node_name
+        self.variant = node.get("variant", node_name.removeprefix(group.name))
+        self.ty = f"crate::{node_name}"
+
+
+groups = [Group(group_name, group) for group_name, group in ast.items()]
+
+# ------------------------------------------------------------------------------
+# Preamble
+
+out = ["""
+// This is a generated file. Don't modify it by hand!
+// Run `crates/ruff_python_ast/generate.py` to re-generate the file.
+"""]  # fmt: skip
+
+# ------------------------------------------------------------------------------
+# Owned enum
+#
+# For each group, we create an enum that contains an owned copy of a syntax
+# node.
```

Comment:
I think each of those blocks could be functions instead where the inline comment becomes a docstring. 

It would also be nice if the docstring could contain. code snippet of what it generates:

Generates the owned enum for each group, e.g `Stmt`, `Expr` etc. For each, generate:

* The enum
* Implement `From<Variant>`
* Implement `Ranged`
*